### PR TITLE
Add suite for CSI tests

### PR DESCRIPTION
**Pull request by jsafrane**
Goal: create test definitions for installed CSI driver.

Usage: `TEST_CSI_DRIVER_FILES=my-csi-driver.yaml openshift-tests run openshift/csi` in a custom prow job.

See  k8s.io/kubernetes/test/e2e/storage/external for yaml file definition and upstream usage

### CODE REVIEW

**Comment by jsafrane** [comment]
@deads2k, FYI.

**Comment by deads2k** [comment]
/hold

I'd like to get closer to master before we mess with test flags.  https://github.com/openshift/origin/pull/23244 has a good shot at passing now.  I've created patches to directly set the global flags that should actually be represented as some kind of options struct with an optional binding.

**Comment by jsafrane** [comment]
@deads2k, I can see #23244 has merged. What is the right way how to pass cmdline flags to kubernetes test suite now? Especially if the flag that I need to set is only in [k8s.io/kubernetes/test/e2e/storage/external](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/external/external.go#L52) and is not available in `framework.RegisterCommonFlags`/`framework.RegisterStorageFlags`/...?

**Comment by smarterclayton** [comment]
One option is to just make an accessor in your package that allows it to be set.

**Comment by jsafrane** [comment]
> One option is to just make an accessor in your package that allows it to be set.

That means every package that defines e2e tests and uses `flag` should also export a public function with similar semantics. There are only two or three such packages, I added a public function to CSI tests upstream: https://github.com/kubernetes/kubernetes/pull/79647 and backported here.

This worked:

```
TEST_CSI_DRIVER_FILES=$PWD/ebs.yaml openshift-tests run openshift/csi/parallel
```

**Comment by jsafrane** [comment]
@smarterclayton, it looks like most of `openshift-tests.go` is your work, PTAL.

File: cmd/openshift-tests/openshift-tests.go [hunk 1]
```diff
@@ -169,6 +174,9 @@ func newRunUpgradeCommand() *cobra.Command {
 				if len(upgradeOpt.ToImage) == 0 {
 					return fmt.Errorf("--to-image must be specified to run an upgrade test")
 				}
+				if err := initCSITests(); err != nil {
```
Comment (smarterclayton) [inline cmd/openshift-tests/openshift-tests.go]: Move this into initProvider, unless there is a reason for it to be separate.

File: cmd/openshift-tests/openshift-tests.go [hunk 2]
```diff
@@ -338,3 +349,18 @@ func decodeProviderTo(provider string, testContext *e2e.TestContextType) error {
 	klog.V(2).Infof("Provider %s: %#v", testContext.Provider, testContext.CloudConfig)
 	return nil
 }
+
+// Initialize openshift/csi/parallel suite, i.e. define CSI tests from TEST_CSI_DRIVER_FILES.
+func initCSITests() error {
+	driverList := os.Getenv("TEST_CSI_DRIVER_FILES")
```
Comment (smarterclayton) [inline cmd/openshift-tests/openshift-tests.go]: Please add a TODO "Replace environment variable with suite specific argument"

File: test/extended/util/test.go [hunk 3]
```diff
@@ -210,7 +210,11 @@ func AnnotateTestSuite() {
 		if !excludedTestsFilter.MatchString(name) {
 			isSerial := strings.Contains(name, "[Serial]")
 			isConformance := strings.Contains(name, "[Conformance]")
+			isCSI := strings.Contains(name, "[CSI]")
 			switch {
+			// There is currently no [Serial] [CSI] test, just to be safe and prevent flakes.
+			case isCSI && !isSerial:
+				node.SetText(node.Text() + " [Suite:openshift/csi/parallel]")
```
Comment (smarterclayton) [inline test/extended/util/test.go]: You don't need the parallel part.  The runner automatically separates out serial and parallel inside a single run.  The reason we have openshift/conformance/parallel and openshift/conformance/serial is so that we can run two different e2e jobs because of the length of time the tests take to run.  For smaller suites, there is no reason to do that.

File: test/extended/util/test.go [hunk 4]
```diff
@@ -401,8 +405,7 @@ var (
 			`\[Driver: iscsi\]`, // https://bugzilla.redhat.com/show_bug.cgi?id=1711627
 			`\[Driver: ceph\]\[Feature:Volumes\] \[Testpattern: Pre-provisioned PV \(default fs\)\] subPath should verify container cannot write to subpath`,
 
-			`\[Driver: aws\] \[Testpattern: Dynamic PV \(default fs\)\] provisioning should access volume from different nodes`, // https://bugzilla.redhat.com/show_bug.cgi?id=1711688
-			`\[Driver: nfs\] \[Testpattern: Dynamic PV \(default fs\)\] provisioning should access volume from different nodes`, // https://bugzilla.redhat.com/show_bug.cgi?id=1711688
+			`\[Testpattern: Dynamic PV \(default fs\)\] provisioning should access volume from different nodes`, // https://bugzilla.redhat.com/show_bug.cgi?id=1711688
```
Comment (smarterclayton) [inline test/extended/util/test.go]: Does this lose valuable tests (i.e. are there now tests not being run because of this)?  Or were NFS and AWS the only ones that ran during the suite anyway and they were both sipped?

Comment (jsafrane) [reply test/extended/util/test.go]: It's a normal test like any other, it does not bring any great value.

Problem with the test on our clusters is that we run each node in a different AWS zone. The test needs at least two nodes in each zone and thus it fails.
We're trying to fix the test upstream: https://github.com/kubernetes/kubernetes/pull/78697
(approval would be nice).

File: cmd/openshift-tests/openshift-tests.go (hunk 1 above)
Comment (jsafrane) [reply cmd/openshift-tests/openshift-tests.go]: moved

File: cmd/openshift-tests/openshift-tests.go (hunk 2 above)
Comment (jsafrane) [reply cmd/openshift-tests/openshift-tests.go]: added

File: test/extended/util/test.go (hunk 3 above)
Comment (jsafrane) [reply test/extended/util/test.go]: removed `parallel` suffix

**Comment by jsafrane** [comment]
@smarterclayton feedback addressed, PTAL.

File: test/extended/util/test.go (hunk 4 above)
Comment (smarterclayton) [reply test/extended/util/test.go]: Wouldn't this pattern block NFS from being tested against vsphere clusters?

Comment (smarterclayton) [reply test/extended/util/test.go]: Can you link to a bug to restore this test?  Someone later on should have an easy reference for why it is disabled.

**Comment by smarterclayton** [comment]
Just one more comment (link to bug in the disable text so we have a breadcrumb in future)

File: test/extended/util/test.go (hunk 4 above)
Comment (jsafrane) [reply test/extended/util/test.go]: Well, there is a link to a bug.

> Wouldn't this pattern block NFS from being tested against vsphere clusters?

Yes, it would. It's IMO better to skip the test everywhere instead of ugly while-listing regexp (in a blacklist!). The test is being fixed and may be even merged before this PR lands, we're just missing approval.

Comment (jsafrane) [reply test/extended/util/test.go]: Now that #23361 is merged, no change in skipped tests is necessary.

**Comment by jsafrane** [comment]
Rebased, @smarterclayton PTAL

File: test/extended/util/test.go [hunk 5]
```diff
@@ -215,7 +215,10 @@ func AnnotateTestSuite() {
 		if !excludedTestsFilter.MatchString(name) {
 			isSerial := strings.Contains(name, "[Serial]")
 			isConformance := strings.Contains(name, "[Conformance]")
+			isCSI := strings.Contains(name, "[CSI]")
```
Comment (smarterclayton) [inline test/extended/util/test.go]: This is redundant (or should be), you should be doing this on like 485 (replace [CSI] with `[Suite:openshift/csi]`

Comment (jsafrane) [reply test/extended/util/test.go]: I believe that if I remove `isCSI`, CSI tests will be added also to `Suite:openshift/conformance/parallel` in the switch below. I still need `isCSI` not to do that.

I'll remove marking the test with `[CSI]`, I agree it's redundant.

Comment (jsafrane) [reply test/extended/util/test.go]: Fixed with `hasSuite` to catch any test that has already been categorized for OpenShift.

File: test/extended/util/test.go [hunk 6]
```diff
@@ -215,7 +215,10 @@ func AnnotateTestSuite() {
 		if !excludedTestsFilter.MatchString(name) {
 			isSerial := strings.Contains(name, "[Serial]")
 			isConformance := strings.Contains(name, "[Conformance]")
+			hasSuite := strings.Contains(name, "[Suite:openshift")
```
Comment (smarterclayton) [inline test/extended/util/test.go]: Tests can be in multiple suites - why did you add this?

**Comment by smarterclayton** [comment]
Are CSI tests just slow?

**Comment by jsafrane** [comment]
> Hrm, unfortunately it looks like the mock volumes are really slow.

What mock volume tests? Are you complaining about e2e-aws-serial flakes? This PR does not add any new tests to ci/prow/e2e-aws-serial nor changes behavior of the existing ones. And the test suite does not even have any csi-mock storage tests.

Please create a separate issue if you think that storage tests are slow. Storage has huge matrix of features x supported storage backends (x in-tree/out-of-tree) and it usually involves non-trivial setup. We can split them into multiple suites. But it's unrelated to this PR.

**Comment by smarterclayton** [comment]
No, we can't merge this as is because this took the e2e-aws run time from ~30m to 1h, and the difference seems to be entirely CSI mock volume tests running.

**Comment by jsafrane** [comment]
Comparing [the last successful e2e-aws from this PR](https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/23242/pull-ci-openshift-origin-master-e2e-aws/11201) and [the last successful e2e-aws from unrelated PR](https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/23373/pull-ci-openshift-origin-master-e2e-aws/11209):

This PR: 1h50m42s
The other PR: 2h19m46s

This PR: all CSI mock tests take 1m15s - 1m46s
The other PR: 53s - 3m18s (!)

The amount of tests is the same, 926 tests. I checked couple of other successful runs, no big changes in CSI mock tests. This PR does not add new tests to e2e-aws and can't influence its speed (how could it? Prow does not set `TEST_CSI_DRIVER_FILES` and I did not change how conformance/parallel is run).

**Comment by smarterclayton** [comment]
If you know why the tests are slow, can we fix that?  Running mock CSI always is good if mock CSI prevents regressions.  If mock CSI is just slow by nature and it can't be fixed, then add a rule that excludes it from the conformance suite under `[Slow]` (which won't prevent it from being part of CSI suite)

**Comment by jsafrane** [comment]
Yes, we can fix that, but again, as a separate issue. This is IMO orthogonal to this PR. There are couple of low hanging fruit, such as Ceph and Gluster tests that are not that important in 4.x. I'd keep CSI included, as it's the most important storage interface now, it has many moving parts and is still under heavy development.

> e2e-aws run time from ~30m to 1h, and the difference seems to be entirely CSI mock volume tests running.

Hmm, there are 8 CSI-mock tests, each takes say 2 minutes, that's 16 minutes when they run serially. What how many tests do we run in parallel? I believe it's more than 8, so CSI mock can add 2 minutes max. The root cause is somewhere else. It could be still storage tests, 403 out of 926 have `[sig-storage]`. Did I mention huge matrix?

**Comment by smarterclayton** [comment]
AWS currently runs 885, so whatever you changed here added ~40 tests.  Normal origin run takes about 30m for tests, while this took almost 1h.

**Comment by jsafrane** [comment]
Where do you get these numbers?

> AWS currently runs 885 so whatever you changed here added ~40 tests

Picking the last successful e2e-aws unrelated to this PR, https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/23440/pull-ci-openshift-origin-master-e2e-aws/11413 has 918 tests, completed in 1h48m.

Re-testing with this PR, e2e-aws has 908 tests, completed in 1h11m.

Thing is, I get different number of tests each time I run them. Still, this PR does not add any new tests nor change their duration. The randomness comes from somewhere else.

> Normal origin run takes about 30m for tests, while this took almost 1h.

I am afraid that times when e2e-aws took 30m are long gone and it's not problem of this PR.

**Comment by smarterclayton** [comment]
The most recent run was 36m for e2e, with 2 retries (which are executed serially).  I'll let this in for now, I can always move the CSI tests into slow if I have to.  And selectively cut out some tests to get our times back down.

/lgtm

**Comment by jsafrane** [comment]
/hold
for rebase

**Comment by jsafrane** [comment]
rebased
/hold cancel
